### ISSUE TITLE
the trait bound `ic_cdk::export::candid::Nat: Serialize` is not satisfied

### ISSUE THREAD

**Issue by lastmjs** [dfinity/candid#331]
This is related to https://github.com/dfinity/candid/issues/324. I think that all Candid types in Rust should implement `serde::Serialize` and `serde::Deserialize`. This will probably also be a problem with `ic_cdk::export::candid::Principal` (turns out it isn't for some reason).

I'm running into this error when using Rust structs with `candid::Nat` fields:

```
error[E0277]: the trait bound `ic_cdk::export::candid::Nat: Serialize` is not satisfied
    --> canisters/canister1/src/lib.rs:31:335
     |
31   | ...ct CanisterStatusResult { cycles : candid :: Nat } struct SERVICE (candid :: Principal) ;
     |                              ^^^^^^ the trait `Serialize` is not implemented for `ic_cdk::export::candid::Nat`
     |
note: required by a bound in `_::_serde::ser::SerializeStruct::serialize_field`
    --> /home/lastmjs/.cargo/registry/src/github.com-1ecc6299db9ec823/serde-1.0.136/src/ser/mod.rs:1899:12
     |
1899 |         T: Serialize;
     |            ^^^^^^^^^ required by this bound in `_::_serde::ser::SerializeStruct::serialize_field`

For more information about this error, try `rustc --explain E0277`.
error: could not compile `canister1` due to previous error
``` 

Can we get this implemented?

**Comment by chenyan-dfinity** [comment]
The standard solution is to define a new type in your application and implement the traits needed there. It's hard to keep up with traits that people needed for all use cases.

Adding `Serialize` to number types is trivial, but for reference types, it's not that easy.

**Comment by lastmjs** [comment]
Why does `ic_cdk::export::candid::Principal` have `Serialize` implemented? I don't think we want to support all traits, but `serde` `Serialize` and `Deserialize` are so fundamental.

**Comment by chenyan-dfinity** [comment]
I'm neutral on what traits should be implemented, but it's also hard for me to decide which traits goes in the library and which don't. This is a common problem with Rust libraries, not specific to this crate.

My concern for adding the `Serialize` trait is that the trait is not needed to support Candid, and if we add it everywhere, it may create the wrong impression that `Serialize` is needed for Candid. Not sure how important is that concern. If adding the trait can unblock you on the short term, I'm happy to do it.

In the long term, I still think adding a config for the `didc` Rust binding to control the generated struct for `nat` type is the right solution.

**Comment by lastmjs** [comment]
Maybe hold off for a bit, I'm in the middle of a major rearchitecture of some type conversions that may solve this problem for me more elegantly and might be along the lines of more idiomatic Rust usage like you're suggesting.

But it does seem inconsistent to me to implement Deserialize and Serialize for every candid type excluding Nat, so not sure if you could explain that.

**Comment by chenyan-dfinity** [comment]
I think `principal` is the only exception here. It's defined in `ic-types`, not in the candid crate. It has `Serialize` trait because it's used in JSON and protobuf, besides Candid.

**Comment by lastmjs** [comment]
> In the long term, I still think adding a config for the didc Rust binding to control the generated struct for nat type is the right solution.

I agree, and I have used your suggestions to overcome my issues. If we can have control over the generated structs, adding the attributes that we need, I think would work great!

**Comment by witterlee** [comment]
@chenyan-dfinity I encountered the same problem, but I think Nat should support Serialize. Nat is the base data type provided by Dfinity, and as a base type, it is too easy to be used to not support serialization, and encounters limitations in many scenarios where serialized data is needed for use. 

For example: when I was designing the token standard, I tried to keep the transaction history by serialization, and the numbers of the transactions were using the Nat type and encountered this problem.
